amethyst_renderer/src/formats/mod.rs

```diff
@@ -9,34 +9,63 @@ use amethyst_assets::{AssetPrefab, Format, PrefabData, PrefabError, ProgressCoun
 use amethyst_core::specs::prelude::Entity;
 use serde::{Deserialize, Serialize};
 
-use {Mesh, Texture};
+use shape::InternalShape;
+use {Mesh, ShapePrefab, Texture};
 
 mod mesh;
 mod mtl;
 mod texture;
 
+/// Internal mesh loading
+///
+/// ### Type parameters:
+///
+/// `V`: Vertex format to use for generated `Mesh`es, must to be one of:
+///     * `Vec<PosTex>`
+///     * `Vec<PosNormTex>`
+///     * `Vec<PosNormTangTex>`
+///     * `ComboMeshCreator`
+/// `M`: `Format` to use for loading `Mesh`es from file
+#[derive(Deserialize, Serialize)]
+pub enum MeshPrefab<V, M>
+where
+    M: Format<Mesh>,
+    M::Options: for<'a> Deserialize<'a> + Serialize,
+{
+    /// Load an asset Mesh from file
+    Asset(AssetPrefab<Mesh, M>),
+    /// Generate a Mesh from basic type
+    Shape(ShapePrefab<V>),
+}
+
 /// `PrefabData` for loading graphics, ie `Mesh` + `Material`
 ///
 /// ### Type parameters:
 ///
+/// `V`: Vertex format to use for generated `Mesh`es, must to be one of:
+///     * `Vec<PosTex>`
+///     * `Vec<PosNormTex>`
+///     * `Vec<PosNormTangTex>`
+///     * `ComboMeshCreator`
 /// - `M`: `Format` to use for loading `Mesh`es from file
 /// - `T`: `Format` to use for loading `Texture`s from file
 #[derive(Deserialize, Serialize)]
-pub struct GraphicsPrefab<M, T>
+pub struct GraphicsPrefab<V, M = ObjFormat, T = TextureFormat>
```

Comment:
I'd like to do something similar with the font format for the font prefab since we only support one format.

Reply:
Right, yeah. There's a `FontFormat` that contain the fonts we have formats for already, we should default to that

Reply:
The only sad part is you can't use type defaults in typedefs :/

Reply:
I can add that to the todo for the ui PR.